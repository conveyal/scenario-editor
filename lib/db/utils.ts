import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollectionResults(obj: unknown) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollectionResults)
  if (typeof obj === 'object') return mapValues(obj, serializeCollectionResults)
  return obj
}
