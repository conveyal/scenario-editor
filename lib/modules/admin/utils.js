import {connect} from 'react-redux'
import {createStructuredSelector} from 'reselect'

import * as actions from './actions'
import * as select from './selectors'

export function fullyConnect(Component) {
  return connect(createStructuredSelector(select), actions)(Component)
}

const add0 = n => (n >= 10 ? n : `0${n}`)

export function msToDuration(ms) {
  const durationS = Math.floor((Date.now() - ms) / 1000)
  const s = durationS % 60
  const m = Math.floor((durationS / 60) % 60)
  const h = Math.floor((durationS / 60 / 60) % 60)

  return `${add0(h)}:${add0(m)}:${add0(s)}`
}

/**
 * Example commit string: "v3.4.0-118-gbe7199c". Take the last section after
 * splitting on the hyphens and remove the "g".
 */
const githubUrl = 'https://github.com/conveyal/r5/commits'
export function createR5Url(workerCommit) {
  // not a release version
  if (workerCommit.indexOf('-') !== -1) {
    const hash = workerCommit.split('-')[2].slice(1) //
    return `${githubUrl}/${hash}`
  }
  return `${githubUrl}/${workerCommit}`
}

export function createWorkerUrl(instanceId, region) {
  return `https://${region}.console.aws.amazon.com/ec2/v2/home?region=${region}#Instances:instanceId=${instanceId};sort=tag:Name`
}

const C1 = '#2389c9'
const C2 = '#36c923'
const C3 = '#aab4bb'

/**
 * Create a progress bar with a linear-gradient.
 */
export function createLinearGradientForJob(c, d, t) {
  const p1 = (c / t) * 100
  const p2 = (d / t) * 100
  return `linear-gradient(to right,
    ${C1} 0%,
    ${C1} ${p1}%,
    ${C2} ${p1}%,
    ${C2} ${p2}%,
    ${C3} ${p2}%,
    ${C3} 100%)`
}
