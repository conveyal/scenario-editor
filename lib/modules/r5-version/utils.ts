import {
  MINIMUM_R5_VERSION,
  VERSION_PARSE_REGEX,
  OLD_VERSION_PARSE_REGEX
} from './constants'

/**
 * Convert an r5 version to a number for comparing
 * NB Assumes no major, minor, patch, or commit version exceeds 100
 */
export function versionToNumber(version?: string): number {
  const versionString = version || ''
  if (versionString.match(/\./g) && versionString.match(/\./g).length == 1) {
    const versionMatch = versionString.match(VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [, major = '0', minor = '0', commit = '0'] = versionMatch

      return (
        parseInt(major) * 10000 + parseInt(minor) * 100 + parseInt(commit) / 100
      )
    }

    return 0
  } else {
    const versionMatch = versionString.match(OLD_VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [
        ,
        major = '0',
        minor = '0',
        patch = '0',
        commit = '0'
      ] = versionMatch

      return (
        parseInt(major) * 10000 +
        parseInt(minor) * 100 +
        parseInt(patch) +
        parseInt(commit) / 100
      )
    }

    return 0
  }
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = (version || '').search(/-/g) !== -1
  return isTestVersion || workerVersionInRange(version, min, max)
}
