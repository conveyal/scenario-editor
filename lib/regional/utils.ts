function getDefaultCutoff(a: CL.RegionalAnalysis): number {
  if (Array.isArray(a.cutoffsMinutes))
    return a.cutoffsMinutes[Math.floor(a.cutoffsMinutes.length / 2)]
  return a.cutoffMinutes
}

function getDefaultDestinationPointSet(a: CL.RegionalAnalysis): string | void {
  const ids = a.destinationPointSetIds
  if (Array.isArray(ids)) return ids[0]
}

function getDefaultPercentile(a: CL.RegionalAnalysis): number {
  if (Array.isArray(a.travelTimePercentiles))
    return a.travelTimePercentiles[
      Math.floor(a.travelTimePercentiles.length / 2)
    ]
  return a.travelTimePercentile
}

/**
 * Choose
 */
export function getDefaultVariants(analysis: CL.RegionalAnalysis) {
  return {
    cutoff: getDefaultCutoff(analysis),
    destinationPointSetId: getDefaultDestinationPointSet(analysis),
    percentile: getDefaultPercentile(analysis)
  }
}

export function getComparisonVariants(comparison: CL.RegionalAnalysis) {
  return {
    comparisonCutoff: getDefaultCutoff(comparison),
    comparisonDestinationPointSetId: getDefaultDestinationPointSet(comparison),
    comparisonPercentile: getDefaultPercentile(comparison)
  }
}

/**
 * Don't compare incomparable analyses.
 */
export function getValidComparisonAnalyses(
  allAnalyses: CL.RegionalAnalysis[],
  activeJobs: CL.RegionalJob[],
  analysis: CL.RegionalAnalysis
) {
  return allAnalyses.filter(
    (c) =>
      activeJobs.findIndex((job) => job.jobId === c._id) === -1 &&
      c.request.originPointSetKey == null &&
      c.zoom === analysis.zoom &&
      c.width === analysis.width &&
      c.height === analysis.height &&
      c.north === analysis.north &&
      c.west === analysis.west
  )
}
