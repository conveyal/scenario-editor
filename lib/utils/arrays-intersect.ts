// Do the arrays intersect?
export default function arraysIntersect<T>(a1: T[], a2: T[]) {
  for (const v of a1) {
    if (a2.indexOf(v) > -1) return true
  }
  return false
}
