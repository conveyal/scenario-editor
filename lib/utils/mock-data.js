import multi from '@conveyal/woonerf/store/multi'
import promise from '@conveyal/woonerf/store/promise'
import {middleware as fetchMiddleware} from '@conveyal/woonerf/fetch'
import configureStore from 'redux-mock-store'
import thunkMiddleware from 'redux-thunk'
import {TextEncoder} from 'text-encoding'

import {responseToSurface} from '../actions/analysis'
import {ADD_TRIP_PATTERN} from '../constants'
import {TRAVEL_TIME_PERCENTILES} from '../constants/analysis'

export const makeMockStore = configureStore([fetchMiddleware, multi, promise, thunkMiddleware])

export const mockSegmentGeom = {
  type: 'LineString',
  coordinates: [
    [
      -122.0246,
      36.9707
    ],
    [
      -122.0279,
      37.049
    ],
    [
      -121.9799,
      37.2299
    ],
    [
      -121.9445,
      37.324
    ],
    [
      -121.936,
      37.353
    ],
    [
      -121.924,
      37.365
    ]
  ]
}

export const mockSegment = {
  type: 'Feature',
  properties: {},
  geometry: mockSegmentGeom
}

const mockStop1 = {
  stop_id: '1',
  stop_lat: 36.9707,
  stop_lon: -122.0246,
  stop_name: 'Santa Cruz'
}

const mockStop2 = {
  stop_id: '2',
  stop_lat: 37.049,
  stop_lon: -122.0279,
  stop_name: 'Scotts Valley'
}

export const mockStops = [mockStop1, mockStop2]

export const mockPattern = {
  geometry: mockSegmentGeom,
  stops: mockStops,
  trips: [
    {
      trip_id: 'abcd',
      start_time: 12345,
      trip_short_name: 'The Express',
      trip_headsign: 'To Downtown',
      duration: 1234
    }
  ]
}

export const mockRoute = {
  id: 'route1',
  patterns: [mockPattern]
}

export const mockFeed = {
  id: '1',
  routes: [mockRoute],
  routesById: {
    route1: mockRoute
  },
  stopsById: {
    '1': mockStop1,
    '2': mockStop2
  },
  stops: [mockStop1, mockStop2]
}

export const mockMapState = {
  allowExtend: true,
  extendFromEnd: true,
  followRoad: true,
  state: 'add-trip-pattern',
  modificationId: '1234'
}

export const mockTimetable = {
  name: 'Test timetable',
  segmentSpeeds: [],
  dwellTime: 10,
  monday: true,
  tuesday: true,
  wednesday: true,
  thursday: true,
  friday: true,
  saturday: false,
  sunday: false,
  headwaySecs: 900, // 15 minutes
  startTime: 28800, // 8am
  endTime: 57600, // 4pm
  patternTrips: ['abcd'],
  sourceTrip: 'abcd'
}

export const mockModification = {
  bidirectional: false,
  entries: [],
  feed: mockFeed.id,
  id: '1234',
  name: 'Test Modification',
  routes: ['route1'],
  segments: [],
  showOnMap: false,
  timetables: [mockTimetable],
  trips: ['abcd'],
  type: ADD_TRIP_PATTERN,
  variants: [0]
}

export const mockProject = {
  bounds: {
    north: 39,
    east: -76,
    south: 38,
    west: -77
  },
  description: 'Project description',
  id: '1',
  name: 'Mock Project'
}

export const mockScenario = {
  bundleId: '1',
  feeds: [mockFeed],
  feedsById: {
    [mockFeed.id]: mockFeed
  },
  id: '1',
  name: 'Mock Scenario',
  projectId: '1',
  variants: ['Default']
}

export const mockStores = {
  init: {
    analysis: {
      isFetchingIsochrone: false,
      isochroneCutoff: 60
    },
    mapState: {
      activeTrips: [],
      center: {
        lat: 38.8886,
        lon: -77.043
      },
      components: [],
      zoom: 12
    },
    network: {
      error: null,
      outstandingRequests: 0
    },
    project: {
      currentProject: null,
      projects: [],
      projectsById: {}
    },
    routing: {
      locationBeforeTransitions: null
    },
    scenario: {
      bundles: [],
      bundlesById: {},
      currentBundle: {
        id: '1'
      },
      currentScenario: mockScenario,
      feeds: [mockFeed],
      feedsById: {
        [mockFeed.id]: mockFeed
      },
      modifications: [mockModification],
      modificationsById: {
        [mockModification.id]: mockModification
      },
      modificationsByType: {
        [mockModification.type]: [mockModification]
      },
      scenarios: [],
      scenariosById: {},
      variants: ['Default']
    },
    user: {
      idToken: null
    }
  }
}

export const mockScenarioApplicationError = {
  'modificationId': 'Add Trip Pattern',
  'title': "Error while applying the modification entitled 'Add Trip Pattern'.",
  'messages': [
    'Headway is not greater than zero.'
  ]
}

export const mockProfileRequest = {
  date: '2016-01-16',
  fromTime: 25200,
  toTime: 32400,
  accessModes: 'WALK',
  directModes: 'WALK',
  egressModes: 'WALK',
  transitModes: 'TRANSIT',
  walkSpeed: 1.3888888888888888,
  bikeSpeed: 4.166666666666667,
  carSpeed: 20,
  streetTime: 90,
  maxWalkTime: 20,
  maxBikeTime: 20,
  maxCarTime: 45,
  minBikeTime: 10,
  minCarTime: 10,
  suboptimalMinutes: 5,
  reachabilityThreshold: 0,
  bikeSafe: 1,
  bikeSlope: 1,
  bikeTime: 1,
  bikeTrafficStress: 4,
  monteCarloDraws: 200,
  maxRides: 4
}

export const mockRegionalAnalyses = [
  {
    zoom: 9,
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 1',
    grid: 'Jobs_total',
    scenarioId: '1234',
    variant: 0,
    bundleId: '5678',
    cutoffMinutes: 60,
    id: 'abcd',
    request: mockProfileRequest,
    // percentile -1 indicates this analysis was done using Andrew Owen-style average instantaneous accessibility
    travelTimePercentile: -1
  },
  {
    zoom: 9,
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 2',
    grid: 'Jobs_total',
    scenarioId: '4321',
    variant: 1,
    bundleId: '5678',
    cutoffMinutes: 60,
    id: 'efgh',
    request: mockProfileRequest,
    travelTimePercentile: 42
  }
]

export const mockGrid = {
  zoom: 9,
  width: 300,
  height: 300,
  north: 20024,
  west: 20015,
  data: [],
  min: 0,
  max: 2000
}

export const mockBookmark = {
  name: 'Bookmark',
  profileRequest: mockProfileRequest,
  isochroneCutoff: 42,
  destinationGrid: 'Workers_total',
  projectId: 'PROJECT'
}

export const mockSurfaceParameters = {
  width: 5,
  height: 5,
  nSamples: 5,
  zoom: 9,
  north: 40000,
  west: 40000
}

// this is a function because responseToSurface is destructive and can lead to test leakage if we
// don't have a brand-new surface each time
export const createMockSurfaceData = () => {
  const {width, height, nSamples, zoom, north, west} = mockSurfaceParameters

  const metadata = {
    scenarioApplicationWarnings: [mockScenarioApplicationError]
  }

  const metadataEncoded = new TextEncoder().encode(JSON.stringify(metadata))
  const bufSize = 9 * 4 + // header
    width * height * nSamples * 4 + // data
    metadataEncoded.length // metadata

  const buffer = new ArrayBuffer(bufSize)
  const surfaceData = new Int32Array(buffer, 0, 9 + width * height * nSamples)

  // text header ("ACCESSGR")
  surfaceData[0] = 1162036033 // ACCE
  surfaceData[1] = 1380406099 // SSGR

  // version
  surfaceData[2] = 0

  surfaceData[3] = zoom
  surfaceData[4] = west
  surfaceData[5] = north
  surfaceData[6] = width
  surfaceData[7] = height
  surfaceData[8] = nSamples

  // phew, header is done
  for (let y = 0, pixel = 9; y < height; y++) {
    for (let x = 0; x < width; x++) {
      let prev = 0
      for (let z = 0; z < nSamples; z++, pixel++) {
        const pixelVal = getMockSurfaceValue(x, y, z)
        surfaceData[pixel] = pixelVal - prev // values are delta-coded within each pixel
        prev = pixelVal
      }
    }
  }

  // copy metadata
  const metadataBuffer = new Uint8Array(buffer, buffer.byteLength - metadataEncoded.byteLength, metadataEncoded.length)
  for (let i = 0; i < metadataEncoded.length; i++) {
    metadataBuffer[i] = metadataEncoded[i]
  }

  return buffer
}

export const mockSurface = responseToSurface(createMockSurfaceData())

// Return the value that should be stored in sample z of pixel x, y of the mock surface
export function getMockSurfaceValue (x, y, z) {
  // this used to simply return z, 2 * z and 3 * z when we had an array of [5, 10, 15 . . . 95]th
  // percentiles. However, when we switched to requesting specific percentiles that stopped working.
  // This translates the position in the access grid into a percentile, then subtracts five and divides
  // by five to get the position in the old array, so that tests written assuming the old structure continue
  // to pass.
  if (x < 2 && y < 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y >= 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y < 2) return 2 * (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x < 2 && y >= 2) return 3 * (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
}

const destinationGridData = new Int32Array(mockSurfaceParameters.width * mockSurfaceParameters.height)
// top half of grid is 1, bottom half is 2
for (let i = 0; i < 3 * mockSurfaceParameters.width; i++) destinationGridData[i] = 1
for (let i = 3 * mockSurfaceParameters.width; i < destinationGridData.length; i++) destinationGridData[i] = 2

export const mockDestinationGrid = {
  ...mockSurfaceParameters,
  west: mockSurfaceParameters.west - 1, // shift slightly so that we are testing misaligned grids
  north: mockSurfaceParameters.north - 1,
  data: destinationGridData
}

export const releaseVersions = [ 'v3.0.0', 'v2.2.0', 'v2.0.0' ]
export const allVersions = [ 'v3.0.0',
  'v2.2.0-2-g789df3e',
  'v2.2.0',
  'v2.0.0-19-g589f32f',
  'v2.0.0' ]
