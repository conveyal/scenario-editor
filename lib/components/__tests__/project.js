// @flow

import React from 'react'
import renderer from 'react-test-renderer'
import {mount} from 'enzyme'
import {mountToJson} from 'enzyme-to-json'

import Project from '../project'
import {mockProject} from '../../utils/mock-data'

describe('Component > Project', () => {
  function createTestData () {
    const deleteProjectFn = jest.fn()
    const downloadScenarioFn = jest.fn()
    const loadFn = jest.fn()
    const setCurrentProjectFn = jest.fn()
    return {
      component: (
        <Project
          _id={mockProject._id}
          children='Children content'
          deleteProject={deleteProjectFn}
          downloadScenario={downloadScenarioFn}
          load={loadFn}
          project={mockProject}
          setCurrentProject={setCurrentProjectFn}
          />
      ),
      deleteProjectFn,
      downloadScenarioFn,
      loadFn,
      setCurrentProjectFn
    }
  }

  afterEach(() => {
    // jsdom persists across test cases, so we need to remove stuff so it doesn't
    // get found again
    const ulTags = document.getElementsByTagName('ul')
    Object.keys(ulTags).forEach(key => {
      ulTags.item(key).remove()
    })
  })

  it('renders correctly', () => {
    const testData = createTestData()
    const tree = renderer
      .create(testData.component)
      .toJSON()
    expect(tree).toMatchSnapshot()
    expect(testData.deleteProjectFn).not.toBeCalled()
    expect(testData.downloadScenarioFn).not.toBeCalled()
    expect(testData.setCurrentProjectFn).not.toBeCalled()
    expect(testData.loadFn).toBeCalled()
  })

  it('should show print select dialogue', () => {
    const testData = createTestData()
    const tree = mount(testData.component)

    tree.find('[name="Print project"]').simulate('click')

    // snapshot to make sure modal is created
    expect(mountToJson(tree)).toMatchSnapshot()

    // find contents of modal
    const ulTag = document.getElementsByTagName('ul')[0]
    expect(ulTag.outerHTML).toMatchSnapshot()
  })

  it('should show export select dialogue', () => {
    const testData = createTestData()
    const tree = mount(testData.component)

    tree.find('[name="Export project"]').simulate('click')

    // snapshot to make sure modal is created
    expect(mountToJson(tree)).toMatchSnapshot()

    // find contents of modal
    const ulTag = document.getElementsByTagName('ul')[0]
    expect(ulTag.outerHTML).toMatchSnapshot()
  })
})
