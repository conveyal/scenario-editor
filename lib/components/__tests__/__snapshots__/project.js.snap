// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component > Project renders correctly 1`] = `
<div>
  <div
    className="ApplicationDockTitle"
  >
    <i
      className="fa fa-cube fa-fw "
    />
     
    Mock Project
    <span
      className="Tip pull-right"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <a
        name="Delete project"
        onClick={[Function]}
        tabIndex={0}
        type="button"
      >
        <i
          className="fa fa-trash fa-fw "
        />
      </a>
    </span>
    <span
      className="Tip pull-right"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <a
        name="Edit project name"
        onClick={[Function]}
        style={Object {}}
        type="button"
      >
        <i
          className="fa fa-gear fa-fw "
        />
      </a>
    </span>
    <span
      className="Tip pull-right"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <a
        name="Import route alignment shapefile"
        onClick={[Function]}
        style={Object {}}
        type="button"
      >
        <i
          className="fa fa-globe fa-fw "
        />
      </a>
    </span>
    <span
      className="Tip pull-right"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <a
        name="Print project"
        onClick={[Function]}
        tabIndex={0}
        type="button"
      >
        <i
          className="fa fa-print fa-fw "
        />
      </a>
    </span>
    <span
      className="Tip pull-right"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <a
        name="Export project"
        onClick={[Function]}
        tabIndex={0}
        type="button"
      >
        <i
          className="fa fa-upload fa-fw "
        />
      </a>
    </span>
  </div>
  <div
    className="InnerDock "
    style={null}
  >
    <div
      className="block"
    >
      Loading GTFS bundle & modificationsâ€¦
    </div>
  </div>
</div>
`;

exports[`Component > Project should show export select dialogue 1`] = `
<Project
  _id="1"
  deleteProject={[Function]}
  downloadScenario={[Function]}
  load={[Function]}
  project={
    Object {
      "_id": "1",
      "accessGroup": "accessGroupName",
      "bundleId": "1",
      "createdAt": "2017-11-08T12:30:50.133Z",
      "createdBy": "person@gmail.com",
      "name": "Mock Project",
      "nonce": "12345",
      "regionId": "1",
      "updatedAt": "2017-11-08T12:31:19.602Z",
      "updatedBy": "person@gmail.com",
      "variants": Array [
        "Default",
      ],
    }
  }
  setCurrentProject={[Function]}
>
  <div>
    <div
      className="ApplicationDockTitle"
    >
      <Icon
        type="cube"
      >
        <i
          className="fa fa-cube fa-fw "
        />
      </Icon>
       
      Mock Project
      <IconLink
        className="pull-right"
        onClick={[Function]}
        title="Delete project"
        type="trash"
      >
        <Link
          className="pull-right"
          onClick={[Function]}
          title="Delete project"
          type="button"
        >
          <Tip
            className="pull-right"
            tip="Delete project"
          >
            <span
              className="Tip pull-right"
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <a
                name="Delete project"
                onClick={[Function]}
                tabIndex={0}
                type="button"
              >
                <Icon
                  type="trash"
                >
                  <i
                    className="fa fa-trash fa-fw "
                  />
                </Icon>
              </a>
            </span>
          </Tip>
        </Link>
      </IconLink>
      <IconLink
        className="pull-right"
        title="Edit project name"
        to="/projects/1/edit"
        type="gear"
      >
        <Link
          className="pull-right"
          title="Edit project name"
          to="/projects/1/edit"
          type="button"
        >
          <Tip
            className="pull-right"
            tip="Edit project name"
          >
            <span
              className="Tip pull-right"
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <Link
                name="Edit project name"
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/projects/1/edit"
                type="button"
              >
                <a
                  name="Edit project name"
                  onClick={[Function]}
                  style={Object {}}
                  type="button"
                >
                  <Icon
                    type="gear"
                  >
                    <i
                      className="fa fa-gear fa-fw "
                    />
                  </Icon>
                </a>
              </Link>
            </span>
          </Tip>
        </Link>
      </IconLink>
      <IconLink
        className="pull-right"
        title="Import route alignment shapefile"
        to="/projects/1/import-shapefile"
        type="globe"
      >
        <Link
          className="pull-right"
          title="Import route alignment shapefile"
          to="/projects/1/import-shapefile"
          type="button"
        >
          <Tip
            className="pull-right"
            tip="Import route alignment shapefile"
          >
            <span
              className="Tip pull-right"
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <Link
                name="Import route alignment shapefile"
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/projects/1/import-shapefile"
                type="button"
              >
                <a
                  name="Import route alignment shapefile"
                  onClick={[Function]}
                  style={Object {}}
                  type="button"
                >
                  <Icon
                    type="globe"
                  >
                    <i
                      className="fa fa-globe fa-fw "
                    />
                  </Icon>
                </a>
              </Link>
            </span>
          </Tip>
        </Link>
      </IconLink>
      <Link
        className="pull-right"
        onClick={[Function]}
        title="Print project"
      >
        <Tip
          className="pull-right"
          tip="Print project"
        >
          <span
            className="Tip pull-right"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            <a
              name="Print project"
              onClick={[Function]}
              tabIndex={0}
              type="button"
            >
              <Icon
                type="print"
              >
                <i
                  className="fa fa-print fa-fw "
                />
              </Icon>
            </a>
          </span>
        </Tip>
      </Link>
      <Link
        className="pull-right"
        onClick={[Function]}
        title="Export project"
      >
        <Tip
          className="pull-right"
          tip="Export project"
        >
          <span
            className="Tip pull-right"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            <a
              name="Export project"
              onClick={[Function]}
              tabIndex={0}
              type="button"
            >
              <Icon
                type="upload"
              >
                <i
                  className="fa fa-upload fa-fw "
                />
              </Icon>
              <SelectScenario
                action={[Function]}
                label="export"
                onHide={[Function]}
                scenarios={
                  Array [
                    "Default",
                  ]
                }
              >
                <Component
                  onRequestClose={[Function]}
                >
                  <Modal
                    ariaHideApp={true}
                    bodyOpenClassName="ReactModal__Body--open"
                    className={
                      Object {
                        "base": "CustomModal",
                      }
                    }
                    closeTimeoutMS={0}
                    contentLabel="Modal"
                    isOpen={true}
                    onRequestClose={[Function]}
                    parentSelector={[Function]}
                    portalClassName="ReactModalPortal"
                    shouldCloseOnOverlayClick={true}
                    style={
                      Object {
                        "overlay": Object {
                          "zIndex": 2500,
                        },
                      }
                    }
                  >
                    <noscript />
                  </Modal>
                </Component>
              </SelectScenario>
            </a>
          </span>
        </Tip>
      </Link>
    </div>
    <InnerDock>
      <div
        className="InnerDock "
        style={
          Object {
            "height": 768,
          }
        }
      >
        <div
          className="block"
        >
          Loading GTFS bundle & modificationsâ€¦
        </div>
      </div>
    </InnerDock>
  </div>
</Project>
`;

exports[`Component > Project should show export select dialogue 2`] = `"<ul><li><a href=\\"\\" tabindex=\\"0\\" target=\\"\\">Default</a></li></ul>"`;

exports[`Component > Project should show print select dialogue 1`] = `
<Project
  _id="1"
  deleteProject={[Function]}
  downloadScenario={[Function]}
  load={[Function]}
  project={
    Object {
      "_id": "1",
      "accessGroup": "accessGroupName",
      "bundleId": "1",
      "createdAt": "2017-11-08T12:30:50.133Z",
      "createdBy": "person@gmail.com",
      "name": "Mock Project",
      "nonce": "12345",
      "regionId": "1",
      "updatedAt": "2017-11-08T12:31:19.602Z",
      "updatedBy": "person@gmail.com",
      "variants": Array [
        "Default",
      ],
    }
  }
  setCurrentProject={[Function]}
>
  <div>
    <div
      className="ApplicationDockTitle"
    >
      <Icon
        type="cube"
      >
        <i
          className="fa fa-cube fa-fw "
        />
      </Icon>
       
      Mock Project
      <IconLink
        className="pull-right"
        onClick={[Function]}
        title="Delete project"
        type="trash"
      >
        <Link
          className="pull-right"
          onClick={[Function]}
          title="Delete project"
          type="button"
        >
          <Tip
            className="pull-right"
            tip="Delete project"
          >
            <span
              className="Tip pull-right"
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <a
                name="Delete project"
                onClick={[Function]}
                tabIndex={0}
                type="button"
              >
                <Icon
                  type="trash"
                >
                  <i
                    className="fa fa-trash fa-fw "
                  />
                </Icon>
              </a>
            </span>
          </Tip>
        </Link>
      </IconLink>
      <IconLink
        className="pull-right"
        title="Edit project name"
        to="/projects/1/edit"
        type="gear"
      >
        <Link
          className="pull-right"
          title="Edit project name"
          to="/projects/1/edit"
          type="button"
        >
          <Tip
            className="pull-right"
            tip="Edit project name"
          >
            <span
              className="Tip pull-right"
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <Link
                name="Edit project name"
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/projects/1/edit"
                type="button"
              >
                <a
                  name="Edit project name"
                  onClick={[Function]}
                  style={Object {}}
                  type="button"
                >
                  <Icon
                    type="gear"
                  >
                    <i
                      className="fa fa-gear fa-fw "
                    />
                  </Icon>
                </a>
              </Link>
            </span>
          </Tip>
        </Link>
      </IconLink>
      <IconLink
        className="pull-right"
        title="Import route alignment shapefile"
        to="/projects/1/import-shapefile"
        type="globe"
      >
        <Link
          className="pull-right"
          title="Import route alignment shapefile"
          to="/projects/1/import-shapefile"
          type="button"
        >
          <Tip
            className="pull-right"
            tip="Import route alignment shapefile"
          >
            <span
              className="Tip pull-right"
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <Link
                name="Import route alignment shapefile"
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/projects/1/import-shapefile"
                type="button"
              >
                <a
                  name="Import route alignment shapefile"
                  onClick={[Function]}
                  style={Object {}}
                  type="button"
                >
                  <Icon
                    type="globe"
                  >
                    <i
                      className="fa fa-globe fa-fw "
                    />
                  </Icon>
                </a>
              </Link>
            </span>
          </Tip>
        </Link>
      </IconLink>
      <Link
        className="pull-right"
        onClick={[Function]}
        title="Print project"
      >
        <Tip
          className="pull-right"
          tip="Print project"
        >
          <span
            className="Tip pull-right"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            <a
              name="Print project"
              onClick={[Function]}
              tabIndex={0}
              type="button"
            >
              <Icon
                type="print"
              >
                <i
                  className="fa fa-print fa-fw "
                />
              </Icon>
              <SelectScenario
                label="print"
                onHide={[Function]}
                projectId="1"
                regionId="1"
                scenarios={
                  Array [
                    "Default",
                  ]
                }
              >
                <Component
                  onRequestClose={[Function]}
                >
                  <Modal
                    ariaHideApp={true}
                    bodyOpenClassName="ReactModal__Body--open"
                    className={
                      Object {
                        "base": "CustomModal",
                      }
                    }
                    closeTimeoutMS={0}
                    contentLabel="Modal"
                    isOpen={true}
                    onRequestClose={[Function]}
                    parentSelector={[Function]}
                    portalClassName="ReactModalPortal"
                    shouldCloseOnOverlayClick={true}
                    style={
                      Object {
                        "overlay": Object {
                          "zIndex": 2500,
                        },
                      }
                    }
                  >
                    <noscript />
                  </Modal>
                </Component>
              </SelectScenario>
            </a>
          </span>
        </Tip>
      </Link>
      <Link
        className="pull-right"
        onClick={[Function]}
        title="Export project"
      >
        <Tip
          className="pull-right"
          tip="Export project"
        >
          <span
            className="Tip pull-right"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            <a
              name="Export project"
              onClick={[Function]}
              tabIndex={0}
              type="button"
            >
              <Icon
                type="upload"
              >
                <i
                  className="fa fa-upload fa-fw "
                />
              </Icon>
            </a>
          </span>
        </Tip>
      </Link>
    </div>
    <InnerDock>
      <div
        className="InnerDock "
        style={
          Object {
            "height": 768,
          }
        }
      >
        <div
          className="block"
        >
          Loading GTFS bundle & modificationsâ€¦
        </div>
      </div>
    </InnerDock>
  </div>
</Project>
`;

exports[`Component > Project should show print select dialogue 2`] = `"<ul><li><a href=\\"/reports/1/projects/1/variants/0\\" tabindex=\\"0\\" target=\\"_blank\\">Default</a></li></ul>"`;
